akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardRegion.scala

```diff
@@ -751,25 +772,50 @@ private[akka] class ShardRegion(
   }
 
   def replyToRegionStatsQuery(ref: ActorRef): Unit = {
-    askAllShards[Shard.ShardStats](Shard.GetShardStats)
-      .map { shardStats =>
-        ShardRegionStats(shardStats.map {
-          case (shardId, stats) => (shardId, stats.entityCount)
-        }.toMap)
+    queryShards[ShardStats](shards, Shard.GetShardStats)
+      .map {
+        case (timeouts, shardStats) =>
+          val qm = ShardingQueries.Metadata(shards.size, timeouts, shardStats)
+          if (qm.unresponsiveNonEmpty) log.warning(qm.toString)
+
+          // TODO add ShardRegionStats field for timeouts (proto): ShardRegion.ShardRegionStats(stats, timeouts.toSet)
+          // Until then
+          val stats = shardStats.map(stats => (stats.shardId, stats.entityCount)).toMap ++ timeouts.map((_, 0))
+          ShardRegionStats(stats)
       }
       .recover {
-        case _: AskTimeoutException => ShardRegionStats(Map.empty)
+        case _: AskTimeoutException =>
+          log.error("{} shards queried but timed out after {}.", shards.size, settings.shardRegionQueryTimeout)
+          ShardRegion.ShardRegionStats(shards.keySet.map(_ -> 0).toMap)
       }
       .pipeTo(ref)
   }
 
-  def askAllShards[T: ClassTag](msg: Any): Future[Seq[(ShardId, T)]] = {
+  /**
+   * Query all or a subset of shards, e.g. unresponsive shards that initially
+   * timed out. If the number of `shards` are less than this.shards.size,
+   * this could be a retry. Returns a partitioned set of any shards that may have not replied within the
+   * timeout and shards did reply, to provide retry on only that subset.
+   *
+   * To check subset unresponsive: {{{ queryShards[T](shards.filterKeys(u.contains), T) }}}
+   */
+  def queryShards[T: ClassTag](shards: Map[ShardRegion.ShardId, ActorRef], msg: Any): Future[(Seq[ShardId], Seq[T])] = {
+
     implicit val timeout: Timeout = settings.shardRegionQueryTimeout
-    Future.sequence(shards.toSeq.map {
-      case (shardId, ref) => (ref ? msg).mapTo[T].map(t => (shardId, t))
-    })
+    Future
+      .sequence(shards.toSeq.map {
+        case (shardId, shard) => askOne[T](shard, msg, shardId, timeout)
+      })
+      .map(ShardingQueries.separateE)
   }
 
+  private def askOne[T: ClassTag](
+      shard: ActorRef,
+      msg: Any,
+      shardId: ShardId,
+      timeout: Timeout): Future[Either[ShardId, T]] =
+    (shard ? msg)(timeout).mapTo[T].map(Right(_)).recover { case _: AskTimeoutException => Left(shardId) }
```

Comment:
```
mapTo[T].transform {
  case Success(t) => Right(t)
  case Failure(_: AskTimeoutexception) => Left(shardId)
}
```

Reply:
Does not work here because we do not have a Try or Throwable.

Reply:
And the idea here is to not lose the failure context of the `ShardId` that timed out, so we would not want to flatten it to a Try[T] vs having either `ShardId` or `T`. I like the idea though.

Reply:
That's the try inside of future you transform with `transform`, it's always there, my suggestion would just replace a map+recover with a single transformation.